packages/query/src/block-processor.ts

```diff
@@ -274,6 +278,10 @@ export class BlockProcessor implements BlockProcessorInterface {
         await this.saveTransactions(compactBlock.height, relevantTx);
       }
 
+      for (const spendableNoteRecord of flush?.newNotes ?? []) {
+        await this.maybeUpsertAuctionWithNoteCommitment(spendableNoteRecord);
+      }
+
```

Comment:
I... really don't like this.

You can see on [lines 238-9 above](https://github.com/penumbra-zone/web/pull/1025/files#diff-ccb0493c986b958cf859e773170457e5ff59a4a2e0d742698492247f6ae872daR238-R239) that we're already iterating over `flush.newNotes`. So why don't I put this call to `this.maybeUpsertAuctionWithNoteCommitment()` inside that earlier `for` loop?

The problem is, we need to [call `this.processTransactions()`](https://github.com/penumbra-zone/web/pull/1025/files#diff-ccb0493c986b958cf859e773170457e5ff59a4a2e0d742698492247f6ae872daR272) _before_ calling `this.maybeUpsertAuctionWithNoteCommitment()`, because `this.processTransactions()` is what saves the auction NFT metadata to the database. `this.maybeUpsertAuctionWithNoteCommitment()` [depends on that auction NFT metadata being saved already](https://github.com/penumbra-zone/web/pull/1025/files#diff-ccb0493c986b958cf859e773170457e5ff59a4a2e0d742698492247f6ae872daR465) to be able to detect whether a given note is for an auction NFT; only then will it save the note's commitment to the `AUCTIONS` table.

"So why not just move `this.processTransactions()` to before the block where we handle `flush.newNotes`?" Because `this.processTransactions()` should only run _after_ we've handled `flush.newNotes`, since [we depend on the result of the flush](https://github.com/penumbra-zone/web/pull/1025/files#diff-ccb0493c986b958cf859e773170457e5ff59a4a2e0d742698492247f6ae872daR253) to determine whether there are transactions to process in the first place. It's a catch-22.

This isn't a problem in core because core isn't going back and forth between Rust and TypeScript like we are. If and when we move the block processor into Rust, I believe this issue will be resolved. In the meantime, I can't think of a better solution — but I'm also tired, so reviewers, feel free to make suggestions.

Reply:
Yep, for sure. Done.